onnxscript/rewriter/cast_constant_of_shape.py

```diff
@@ -12,58 +11,34 @@
 logger = logging.getLogger(__name__)
 
 
-def cast_constant_of_shape(
-    shape: Sequence[int],
-    t: Any,
-    dtype: int,
-    match_bindings: dict[str, ir.Value | Any] | None = None,
-) -> pattern.OpPattern:
-    constant = op.ConstantOfShape(shape, value=t)
+def cast_constant_of_shape(shape, scalar, dtype):
+    constant = op.ConstantOfShape(shape, value=scalar)
     return op.Cast(constant, to=dtype)
 
 
-def fused_cast_constant_of_shape(
-    shape: Sequence[int], t: Any, dtype: int, match_bindings: dict[str, ir.Value | Any]
-) -> pattern.OpPattern:
-    del dtype  # unused
-    del t  # unused
-    v_dtype = match_bindings["dtype"]
-    v_t = match_bindings["t"]
-    v_dtype = ir.DataType(v_dtype.value).numpy()  # type: ignore[union-attr]
-    casted_val = ir.Tensor(v_t.value.numpy().astype(v_dtype))  # type: ignore[union-attr]
-    return op.ConstantOfShape(shape, value=casted_val)
-
-
-def cast_constant_of_shape_without_value(
-    shape: Sequence[int],
-    dtype: int,
-    match_bindings: dict[str, ir.Value | Any] | None = None,
-) -> pattern.OpPattern:
-    del match_bindings  # Unused
+def fused_cast_constant_of_shape(op, shape: ir.Value, scalar: ir.Attr, dtype: ir.Attr, **_):
+    # Cast scalar (a TensorProto attribute) to the specified dtype
+    scalar_value = scalar.value.numpy().item()
+    cast_value = onnx.helper.make_tensor("value", dtype.value, (), [scalar_value])
```

Comment:
```suggestion
    cast_value = ir.Tensor(scalar.value.numpy().astype(ir.DataType(dtype.value).numpy()))
```

I think we can do this

Reply:
And we can probably make `ir.DataType(dtype.value).numpy())` a helper function

Reply:
A couple of questions/comments:
* The API should allow a TensorProto created in any form (as it does, as seen here, I think).
* From a readability standpoint, the make_tensor seems easier to read (though, admittedly, it may be because of greater familiarity with that long-used API)
* What happens to types that are not supported in numpy()? Just wondering

Reply:
Right - the api does support tensor protos being created. I just felt it would be a good idea to avoid moving protos around, from a consistency point of view. But it is an issue when the types are not supported by numpy.

I think eventually we will need to replicate some logic from make_tensor to the Tensor class.

Reply:
In terms of readability, we can create the same interface w/o the need for serialization to bytes.

Reply:
I can do this in a separate PR